File: tui/packages/core/src/issues-client.ts
import type {
	Issue,
	IssueComment,
	IssueCommentListResult,
	IssueListResult,
	IssueScope,
} from "@devenv/types";
import type { ClientDeps } from "./client-types";
import { handleFetchError } from "./error-handler";

/**
 * Get issues for a specific app with scope filter and pagination.
 */
export async function getIssues(
	deps: ClientDeps,
	appIdent: string,
	scope: IssueScope = "all",
	sourceType?: string,
	page: number = 1,
	perPage: number = 50,
	search?: string,
	state?: string,
	sort?: string,
	direction?: "asc" | "desc",
	labels?: string[],
): Promise<IssueListResult> {
	const params = new URLSearchParams({
		appIdent,
		scope,
	});
	// Always send state param so server knows the filter (default: "open" for GitHub, "opened" for GitLab)
	params.append("state", state || (sourceType === "github" ? "open" : "opened"));
	if (search) {
		params.append("search", search);
	}
	if (sort) params.append("sort", sort);
	if (direction) params.append("direction", direction);
	if (labels?.length) params.append("labels", labels.join(","));
	params.append("page", String(page));
	params.append("perPage", String(perPage));

	const endpoint = sourceType === "github" ? "github/issues" : "gitlab/issues";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${endpoint}?${params}`,
	);

	if (response.status === 404 || response.status === 400) {
		return {
			items: [],
			totalCount: -1,
			totalPages: -1,
			currentPage: 1,
			perPage,
		};
	}

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as IssueListResult;
}

/**
 * Get a single issue by number.
 */
export async function getIssue(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType?: string,
): Promise<Issue> {
	const endpoint = sourceType === "github" ? "github/issue" : "gitlab/issue";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${endpoint}?appIdent=${encodeURIComponent(appIdent)}&number=${number}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as Issue;
}

/**
 * Get comments for a specific issue.
 */
export async function getIssueComments(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType?: string,
): Promise<IssueCommentListResult> {
	const endpoint =
		sourceType === "github" ? "github/issue-comments" : "gitlab/issue-comments";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${endpoint}?appIdent=${encodeURIComponent(appIdent)}&number=${number}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as IssueCommentListResult;
}

// ─── Mutation Functions ───────────────────────────────────────────────────────

async function issueMutation(
	deps: ClientDeps,
	endpoint: string,
	appIdent: string,
	number: number | undefined,
	body: unknown,
	method: string = "POST",
): Promise<Issue> {
	let url = `${deps.baseUrl}/api/${endpoint}?appIdent=${encodeURIComponent(appIdent)}`;
	if (number !== undefined) {
		url += `&number=${number}`;
	}

	const response = await deps.fetchFn(url, {
		method,
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as Issue;
}

/**
 * Close an issue with an optional reason.
 * reason: "completed" | "not_planned" | "" (empty for no reason)
 */
export async function closeIssue(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType: string | undefined,
	reason?: string,
): Promise<Issue> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	return issueMutation(deps, `${prefix}/issues/close`, appIdent, number, {
		reason: reason ?? "",
	});
}

/**
 * Reopen an issue.
 */
export async function reopenIssue(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType: string | undefined,
): Promise<Issue> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	return issueMutation(deps, `${prefix}/issues/reopen`, appIdent, number, {});
}

/**
 * Set labels on an issue.
 */
export async function setIssueLabels(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType: string | undefined,
	labels: string[],
): Promise<Issue> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	return issueMutation(deps, `${prefix}/issues/labels`, appIdent, number, {
		labels,
	});
}

/**
 * Set assignee on an issue.
 */
export async function setIssueAssignee(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType: string | undefined,
	assignee: string,
): Promise<Issue> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	return issueMutation(deps, `${prefix}/issues/assignee`, appIdent, number, {
		assignee,
	});
}

/**
 * Remove assignee from an issue.
 */
export async function removeIssueAssignee(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType: string | undefined,
): Promise<Issue> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	return issueMutation(deps, `${prefix}/issues/unassign`, appIdent, number, {});
}

/**
 * Get all labels for a repository.
 */
export async function getRepoLabels(
	deps: ClientDeps,
	appIdent: string,
	sourceType: string | undefined,
): Promise<string[]> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${prefix}/labels?appIdent=${encodeURIComponent(appIdent)}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	const data = (await response.json()) as { labels: string[] };
	return data.labels;
}

/**
 * Add a comment to an issue.
 */
export async function addIssueComment(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType: string | undefined,
	body: string,
): Promise<IssueComment> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${prefix}/issues/comment?appIdent=${encodeURIComponent(appIdent)}&number=${number}`,
		{
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({ body }),
		},
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as IssueComment;
}

/**
 * Get linked change requests for an issue.
 */
export async function getIssueLinkedCRs(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType?: string,
): Promise<import("@devenv/types").ChangeRequest[]> {
	const endpoint =
		sourceType === "github"
			? "github/issues/linked-crs"
			: "gitlab/issues/linked-crs";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${endpoint}?appIdent=${encodeURIComponent(appIdent)}&number=${number}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as import("@devenv/types").ChangeRequest[];
}

/**
 * Get issues referenced in an issue's body.
 */
export async function getIssueReferencedIssues(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType?: string,
): Promise<import("@devenv/types").Issue[]> {
	const endpoint =
		sourceType === "github"
			? "github/issues/references"
			: "gitlab/issues/references";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${endpoint}?appIdent=${encodeURIComponent(appIdent)}&number=${number}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as import("@devenv/types").Issue[];
}

/**
 * Get linked issues for a change request.
 */
export async function getCRLinkedIssues(
	deps: ClientDeps,
	appIdent: string,
	number: number,
	sourceType?: string,
): Promise<import("@devenv/types").Issue[]> {
	const endpoint =
		sourceType === "github"
			? "github/cr/linked-issues"
			: "gitlab/cr/linked-issues";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${endpoint}?appIdent=${encodeURIComponent(appIdent)}&number=${number}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	return (await response.json()) as import("@devenv/types").Issue[];
}

/**
 * Get all collaborators for a repository.
 */
export async function getRepoCollaborators(
	deps: ClientDeps,
	appIdent: string,
	sourceType: string | undefined,
): Promise<string[]> {
	const prefix = sourceType === "github" ? "github" : "gitlab";
	const response = await deps.fetchFn(
		`${deps.baseUrl}/api/${prefix}/collaborators?appIdent=${encodeURIComponent(appIdent)}`,
	);

	if (!response.ok) {
		await handleFetchError(response, deps.onError);
	}

	const data = (await response.json()) as { collaborators: string[] };
	return data.collaborators;
}

File: server/pkg/gitlab/issues_client.go
package gitlab

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"github.com/friendsfriend/devenv/pkg/issues"
)

// --- GitLab API raw types for issues ---

type glIssue struct {
	ID          int        `json:"id"`
	IID         int        `json:"iid"`
	Title       string     `json:"title"`
	Description string     `json:"description"`
	State       string     `json:"state"`
	WebURL      string     `json:"web_url"`
	Author      NoteAuthor `json:"author"`
	Labels      []string   `json:"labels"`
	Assignees   []struct {
		Name     string `json:"name"`
		Username string `json:"username"`
	} `json:"assignees"`
	Milestone *struct {
		Title string `json:"title"`
	} `json:"milestone,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type glIssueNote struct {
	ID        int        `json:"id"`
	Body      string     `json:"body"`
	Author    NoteAuthor `json:"author"`
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	System    bool       `json:"system"`
}

// --- Conversion helpers ---

func convertGLIssue(issue glIssue) issues.Issue {
	result := issues.Issue{
		ID:          issue.ID,
		IID:         issue.IID,
		Title:       issue.Title,
		Description: issue.Description,
		State:       issue.State,
		WebURL:      issue.WebURL,
		CreatedAt:   issue.CreatedAt,
		UpdatedAt:   issue.UpdatedAt,
		Labels:      issue.Labels,
		Assignees:   issue.Assignees,
		Milestone:   issue.Milestone,
	}
	result.Author.Name = issue.Author.Name
	result.Author.Username = issue.Author.Username
	return result
}

func convertGLIssueNote(note glIssueNote) issues.IssueComment {
	result := issues.IssueComment{
		ID:        note.ID,
		Body:      note.Body,
		CreatedAt: note.CreatedAt,
		UpdatedAt: note.UpdatedAt,
		System:    note.System,
	}
	result.Author.Name = note.Author.Name
	result.Author.Username = note.Author.Username
	return result
}

// --- issues.Client implementation ---

// IssuesClient wraps a GitLab client to implement issues.Client.
type IssuesClient struct {
	c       *client
	project *ProjectInfo
}

// NewIssuesClient creates a new IssuesClient for the given project.
func NewIssuesClient(c Client, project *ProjectInfo) issues.Client {
	glClient := c.(*client)
	return &IssuesClient{
		c:       glClient,
		project: project,
	}
}

func (ic *IssuesClient) projectToIssueRepo() *issues.RepoInfo {
	return &issues.RepoInfo{
		Host:      ic.project.Host,
		Namespace: ic.project.Namespace,
		Project:   ic.project.Project,
	}
}

func (ic *IssuesClient) doGet(url string) ([]byte, int, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("PRIVATE-TOKEN", ic.c.token)
	req.Header.Set("User-Agent", "devenv-cli")
	req.Header.Set("Accept", "application/json")

	resp, err := ic.c.httpClient.Do(req)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to read response: %w", err)
	}
	return body, resp.StatusCode, nil
}

// doGetWithHeaders is like doGet but also returns response headers.
func (ic *IssuesClient) doGetWithHeaders(url string) ([]byte, int, http.Header, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, 0, nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("PRIVATE-TOKEN", ic.c.token)
	req.Header.Set("User-Agent", "devenv-cli")
	req.Header.Set("Accept", "application/json")

	resp, err := ic.c.httpClient.Do(req)
	if err != nil {
		return nil, 0, nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, nil, fmt.Errorf("failed to read response: %w", err)
	}
	return body, resp.StatusCode, resp.Header, nil
}

// GetIssues implements issues.Client.GetIssues.
func (ic *IssuesClient) GetIssues(info *issues.RepoInfo, options *issues.IssueListOptions) (*issues.IssueListResult, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{
			Host:      info.Host,
			Namespace: info.Namespace,
			Project:   info.Project,
		}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))

	page := 1
	perPage := 50
	scope := ""
	search := ""
	state := "opened"
	if options != nil {
		if options.Page > 0 {
			page = options.Page
		}
		if options.PerPage > 0 {
			perPage = options.PerPage
		}
		scope = options.Scope
		search = options.Search
		if options.State != "" {
			state = options.State
		}
	}

	params := url.Values{}
	params.Set("state", state)
	params.Set("page", fmt.Sprintf("%d", page))
	params.Set("per_page", fmt.Sprintf("%d", perPage))
	params.Set("sort", "updated_desc")

	// Map scope to GitLab API params.
	// GitLab uses underscore-separated values for the scope parameter.
	switch scope {
	case "all":
		params.Set("scope", "all")
	case "assigned-to-me":
		params.Set("scope", "assigned_to_me")
	case "created-by-me":
		params.Set("scope", "created_by_me")
	case "no-assignee":
		params.Set("assignee_id", "None")
	}

	if search != "" {
		params.Set("search", search)
	}

	// GitLab state values: "opened", "closed", "all"
	glState := state
	if glState == "open" {
		glState = "opened"
	}
	params.Set("state", glState)

	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues?%s",
		ic.c.baseURL, projectPath, params.Encode())

	body, statusCode, headers, err := ic.doGetWithHeaders(apiURL)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(body))
	}

	// Parse pagination headers
	totalCount := -1
	totalPages := -1
	currentPage := page

	if totalStr := headers.Get("X-Total"); totalStr != "" {
		if t, err := strconv.Atoi(totalStr); err == nil {
			totalCount = t
		}
	}
	if totalPagesStr := headers.Get("X-Total-Pages"); totalPagesStr != "" {
		if tp, err := strconv.Atoi(totalPagesStr); err == nil {
			totalPages = tp
		}
	}
	if pageStr := headers.Get("X-Page"); pageStr != "" {
		if p, err := strconv.Atoi(pageStr); err == nil {
			currentPage = p
		}
	}

	var glIssues []glIssue
	if err := json.Unmarshal(body, &glIssues); err != nil {
		return nil, fmt.Errorf("failed to parse issues: %w", err)
	}

	results := make([]issues.Issue, 0, len(glIssues))
	for _, iss := range glIssues {
		results = append(results, convertGLIssue(iss))
	}

	return &issues.IssueListResult{
		Issues:      results,
		TotalCount:  totalCount,
		TotalPages:  totalPages,
		CurrentPage: currentPage,
		PerPage:     perPage,
	}, nil
}

// GetIssue implements issues.Client.GetIssue.
func (ic *IssuesClient) GetIssue(info *issues.RepoInfo, number int) (*issues.Issue, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{
			Host:      info.Host,
			Namespace: info.Namespace,
			Project:   info.Project,
		}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d",
		ic.c.baseURL, projectPath, number)

	body, statusCode, err := ic.doGet(apiURL)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(body))
	}

	var glIss glIssue
	if err := json.Unmarshal(body, &glIss); err != nil {
		return nil, fmt.Errorf("failed to parse issue: %w", err)
	}

	result := convertGLIssue(glIss)
	return &result, nil
}

// GetIssueComments implements issues.Client.GetIssueComments.
func (ic *IssuesClient) GetIssueComments(info *issues.RepoInfo, number int) (*issues.IssueCommentListResult, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{
			Host:      info.Host,
			Namespace: info.Namespace,
			Project:   info.Project,
		}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d/notes?per_page=100",
		ic.c.baseURL, projectPath, number)

	body, statusCode, err := ic.doGet(apiURL)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(body))
	}

	var glNotes []glIssueNote
	if err := json.Unmarshal(body, &glNotes); err != nil {
		return nil, fmt.Errorf("failed to parse issue notes: %w", err)
	}

	results := make([]issues.IssueComment, 0, len(glNotes))
	for _, n := range glNotes {
		results = append(results, convertGLIssueNote(n))
	}

	return &issues.IssueCommentListResult{
		Comments:    results,
		TotalCount:  -1,
		TotalPages:  -1,
		CurrentPage: 1,
		PerPage:     100,
	}, nil
}

// --- HTTP helpers for mutations ---

func (ic *IssuesClient) doPost(urlStr string, body []byte) ([]byte, int, error) {
	req, err := http.NewRequest("POST", urlStr, bytes.NewReader(body))
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("PRIVATE-TOKEN", ic.c.token)
	req.Header.Set("User-Agent", "devenv-cli")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")

	resp, err := ic.c.httpClient.Do(req)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to read response: %w", err)
	}
	return respBody, resp.StatusCode, nil
}

func (ic *IssuesClient) doPut(urlStr string, body []byte) ([]byte, int, error) {
	req, err := http.NewRequest("PUT", urlStr, bytes.NewReader(body))
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("PRIVATE-TOKEN", ic.c.token)
	req.Header.Set("User-Agent", "devenv-cli")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")

	resp, err := ic.c.httpClient.Do(req)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to read response: %w", err)
	}
	return respBody, resp.StatusCode, nil
}

func (ic *IssuesClient) doDelete(urlStr string, body []byte) ([]byte, int, error) {
	req, err := http.NewRequest("DELETE", urlStr, bytes.NewReader(body))
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("PRIVATE-TOKEN", ic.c.token)
	req.Header.Set("User-Agent", "devenv-cli")
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := ic.c.httpClient.Do(req)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to read response: %w", err)
	}
	return respBody, resp.StatusCode, nil
}

// --- Label name→ID cache helper ---

// labelNameToID fetches all project labels and returns a map of label name→ID.
func (ic *IssuesClient) labelNameToID(proj *ProjectInfo) (map[string]int, error) {
	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/labels?per_page=100", ic.c.baseURL, projectPath)

	body, statusCode, err := ic.doGet(apiURL)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error fetching labels (status %d): %s", statusCode, string(body))
	}

	var glLabels []struct {
		ID   int    `json:"id"`
		Name string `json:"name"`
	}
	if err := json.Unmarshal(body, &glLabels); err != nil {
		return nil, fmt.Errorf("failed to parse labels: %w", err)
	}

	result := make(map[string]int, len(glLabels))
	for _, l := range glLabels {
		result[l.Name] = l.ID
	}
	return result, nil
}

// --- Mutation implementations ---

// CloseIssue implements issues.Client.CloseIssue for GitLab.
func (ic *IssuesClient) CloseIssue(info *issues.RepoInfo, number int, reason string) (*issues.Issue, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d", ic.c.baseURL, projectPath, number)

	payload := map[string]string{"state_event": "close"}
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	respBody, statusCode, err := ic.doPut(apiURL, jsonPayload)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(respBody))
	}

	var glIss glIssue
	if err := json.Unmarshal(respBody, &glIss); err != nil {
		return nil, fmt.Errorf("failed to parse closed issue: %w", err)
	}

	result := convertGLIssue(glIss)
	return &result, nil
}

// ReopenIssue implements issues.Client.ReopenIssue for GitLab.
func (ic *IssuesClient) ReopenIssue(info *issues.RepoInfo, number int) (*issues.Issue, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d", ic.c.baseURL, projectPath, number)

	payload := map[string]string{"state_event": "reopen"}
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	respBody, statusCode, err := ic.doPut(apiURL, jsonPayload)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(respBody))
	}

	var glIss glIssue
	if err := json.Unmarshal(respBody, &glIss); err != nil {
		return nil, fmt.Errorf("failed to parse reopened issue: %w", err)
	}

	result := convertGLIssue(glIss)
	return &result, nil
}

// SetLabels implements issues.Client.SetLabels for GitLab.
func (ic *IssuesClient) SetLabels(info *issues.RepoInfo, number int, labels []string) (*issues.Issue, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d", ic.c.baseURL, projectPath, number)

	payload := map[string]interface{}{
		"labels": labels,
	}
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	respBody, statusCode, err := ic.doPut(apiURL, jsonPayload)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(respBody))
	}

	var glIss glIssue
	if err := json.Unmarshal(respBody, &glIss); err != nil {
		return nil, fmt.Errorf("failed to parse updated issue: %w", err)
	}

	result := convertGLIssue(glIss)
	return &result, nil
}

// AddAssignee implements issues.Client.AddAssignee for GitLab.
func (ic *IssuesClient) AddAssignee(info *issues.RepoInfo, number int, assignee string) (*issues.Issue, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d", ic.c.baseURL, projectPath, number)

	payload := map[string]interface{}{
		"assignee_ids": []string{assignee},
	}
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	respBody, statusCode, err := ic.doPut(apiURL, jsonPayload)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(respBody))
	}

	var glIss glIssue
	if err := json.Unmarshal(respBody, &glIss); err != nil {
		return nil, fmt.Errorf("failed to parse updated issue: %w", err)
	}

	result := convertGLIssue(glIss)
	return &result, nil
}

// RemoveAssignee implements issues.Client.RemoveAssignee for GitLab.
func (ic *IssuesClient) RemoveAssignee(info *issues.RepoInfo, number int) (*issues.Issue, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d", ic.c.baseURL, projectPath, number)

	payload := map[string]interface{}{
		"assignee_ids": []int{},
	}
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	respBody, statusCode, err := ic.doPut(apiURL, jsonPayload)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(respBody))
	}

	var glIss glIssue
	if err := json.Unmarshal(respBody, &glIss); err != nil {
		return nil, fmt.Errorf("failed to parse updated issue: %w", err)
	}

	result := convertGLIssue(glIss)
	return &result, nil
}

// GetRepoLabels implements issues.Client.GetRepoLabels for GitLab.
func (ic *IssuesClient) GetRepoLabels(info *issues.RepoInfo) ([]string, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/labels?per_page=100", ic.c.baseURL, projectPath)

	body, statusCode, err := ic.doGet(apiURL)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(body))
	}

	var glLabels []struct {
		Name string `json:"name"`
	}
	if err := json.Unmarshal(body, &glLabels); err != nil {
		return nil, fmt.Errorf("failed to parse labels: %w", err)
	}

	result := make([]string, 0, len(glLabels))
	for _, l := range glLabels {
		result = append(result, l.Name)
	}
	return result, nil
}

// GetRepoCollaborators implements issues.Client.GetRepoCollaborators for GitLab.
func (ic *IssuesClient) GetRepoCollaborators(info *issues.RepoInfo) ([]string, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/members?per_page=100", ic.c.baseURL, projectPath)

	body, statusCode, err := ic.doGet(apiURL)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusOK {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(body))
	}

	var members []struct {
		Name     string `json:"name"`
		Username string `json:"username"`
	}
	if err := json.Unmarshal(body, &members); err != nil {
		return nil, fmt.Errorf("failed to parse members: %w", err)
	}

	result := make([]string, 0, len(members))
	for _, m := range members {
		result = append(result, m.Username)
	}
	return result, nil
}

// AddComment implements issues.Client.AddComment for GitLab.
func (ic *IssuesClient) AddComment(info *issues.RepoInfo, number int, body string) (*issues.IssueComment, error) {
	proj := ic.project
	if info != nil && info.Namespace != "" && info.Project != "" {
		proj = &ProjectInfo{Host: info.Host, Namespace: info.Namespace, Project: info.Project}
	}

	projectPath := url.QueryEscape(fmt.Sprintf("%s/%s", proj.Namespace, proj.Project))
	apiURL := fmt.Sprintf("%s/api/v4/projects/%s/issues/%d/notes", ic.c.baseURL, projectPath, number)

	payload := map[string]string{"body": body}
	jsonPayload, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	respBody, statusCode, err := ic.doPost(apiURL, jsonPayload)
	if err != nil {
		return nil, err
	}

	if statusCode != http.StatusCreated {
		return nil, fmt.Errorf("GitLab API error (status %d): %s", statusCode, string(respBody))
	}

	var glNote glIssueNote
	if err := json.Unmarshal(respBody, &glNote); err != nil {
		return nil, fmt.Errorf("failed to parse created note: %w", err)
	}

	result := convertGLIssueNote(glNote)
	return &result, nil
}

var _ = log.Printf
